src/libs/SuffixUkkonenTree/index.ts

```diff
@@ -0,0 +1,183 @@
+/* eslint-disable no-continue */
+import {ALPHABET_SIZE, DELIMITER_CHAR_CODE, END_CHAR_CODE, SPECIAL_CHAR_CODE, stringToNumeric} from './utils';
+
+/**
+ * This implements a suffix tree using Ukkonen's algorithm.
+ * A good visualization to learn about the algorithm can be found here: https://brenden.github.io/ukkonen-animation/
+ * Note: This implementation is optimized for performance, not necessarily for readability.
+ *
+ * You probably don't want to use this directly, but rather use @libs/FastSearch.ts as a easy to use wrapper around this.
+ */
+
+/**
+ * Creates a new tree instance that can be used to build a suffix tree and search in it.
+ * The input is a numeric representation of the search string, which can be create using {@link stringToNumeric}.
+ * Separate search values must be separated by the {@link DELIMITER_CHAR_CODE}. The search string must end with the {@link END_CHAR_CODE}.
+ *
+ * The tree will be built using the Ukkonen's algorithm: https://www.cs.helsinki.fi/u/ukkonen/SuffixT1withFigs.pdf
+ */
+function makeTree(numericSearchValues: Int8Array) {
+    const maxNodes = 2 * numericSearchValues.length;
+    // Allocate an ArrayBuffer to store all transitions (flat buffer), 4 bytes per transition (Uint32)
```

Comment:
Here's my attempt at explaining this variable better (but still trying to understanding all this)

```suggestion
    /* 
       This array represents all internal nodes in the suffix tree.
       When building this tree, we'll be given a character in the string, and we need to be able to lookup in constant time
       if there's any edge connected to a node starting with that character. For example, given a tree like this:

               root
              / | \
             a  b  c
 
       and the next character in our string is 'd', we need to be able do check if any of the edges from the root node
       start with the letter 'd', without looping through all the edges.

       To accomplish this, each node gets an array matching the alphabet size (x4 because it's 4 bytes per character).
       So you can imagine if our alphabet was just [a,b,c,d], then each node would get an array like [0,0,0,0].
       If we add an edge starting with 'a', then the root node would be [1,0,0,0]
       So given an arbitrary letter such as 'd', then we can take the position of that letter in its alphabet (position 3 in our example)
       and check whether that index in the array is 0 or 1. If it's a 1, then there's an edge starting with the letter 'd'.

       Note that for efficiency, all nodes are stored in a single flag array. That's how we end up with (maxNodes * alphabet_size * 4).
       In the example of a 4-character alphabet, we'd have an array like this:

          root       root.left     root.right              last possible node
        /     \       /     \       /     \                      /     \
       [0,0,0,0,      0,0,0,0,      0,0,0,0,  .................  0,0,0,0]
     */
```

Does that make sense? I know it's a lot of work, but that level of detail will be very valuable to make this code easier to understand imo.

Reply:
Sounds good. Normally we would store them in an array of maps. transition: Map<char, Node>[]. But to improve performance we first moved to Array of Arrays. transition: Number[][]. Then when we realised that ArrayBuffers are much faster (and they let as parallel the work as shared buffers can shared between different worklets runtimes) We switched to one dimension.

Reply:
No, the sharing between worklet runtimes isn't soemthing thats happening here but something we are very eager to try soon!

Reply:
Very nice explanation by the way Rory, adding that!